packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,221 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+const nodeTypesAreEqual = <T extends TSESTree.Node>(
+  actualTypeNode: TSESTree.Node,
+  expectedTypeNode: T,
+): actualTypeNode is T => {
+  return actualTypeNode.type === expectedTypeNode.type;
+};
+
+const constituentsAreEqual = (
```

Comment:
@sajikix I just realized I might not have very clear in my suggestion about using the type checker. I'm essentially proposing that we check whether the `ts.Type` for a constituent node is `===` any other previous constituent node.

```ts
const parserServices = util.getParserServices(context);
const checker = parserServices.program.getTypeChecker();

const type = checker.getTypeAtLocation(parserServices.parserServices.esTreeNodeToTSNodeMap.get(node));
```

It's a little silly to use `===`, but we're guaranteed[^1] that the TypeScript type checking APIs will use the same type object from two nodes that have the same type.

[^1]: I don't know that this has been explicitly documented anywhere, but a lot of TypeScript tools work on this assumption. It's now one of those things that TypeScript just has to do.